### ISSUE TITLE
Out of memory error trying to boot CURRENT on BeagleBone (White)

### ISSUE THREAD

**Issue by tcort** [Stichting-MINIX-Research-Foundation/minix#104]
I'm having trouble booting my BeagleBone (White). I'm getting out of memory errors.
## bisect results

I tested a bunch of commits and determined that the boot failure was introduced with commit 0a6a1f1d05b60e214de2f05a7310ddd1f0e590e7

| good/bad | hash | shortlog |
| --- | --- | --- |
| bad | 1d80a87694cdb69a2cb41223306d4bbba5289571 | bootloader: do not try to autoload NetBSD modules |
| bad | bcc17a8117948424aaf38a46f0a761e165407e79 | Integrate magic instrumentation into build system |
| bad | 0ce7112c0b841a8aa0247b15b32690c48d29cf57 | flist: add bitcode entries for crypto libraries |
| bad | 0a6a1f1d05b60e214de2f05a7310ddd1f0e590e7 | NetBSD re-synchronization of the source tree |
| good | 8933525b85c631b3d20bed940561520be2b81891 | Fix default partition in clientctl. |
| good | 2e23f1755e14162eb3fa2237285dec556ccd360f | libc: fix needless malloc failures |
| good | 41a30155b93b5fdc3b6a018df059658fbb97e8b4 | Fix gcc -O1 builds for ARM |
## What I Did

I grabbed the latest sources (commit 1d80a87694cdb69a2cb41223306d4bbba5289571). I also made two minor changes to the `bmp085` temperature sensor driver (removing unused variables) but that shouldn't have caused this issue. I built a Minix/arm image for the BeagleBone on my Debian/amd64 8.0 (jessie) machine by setting `U_BOOT_BIN_DIR=build/am335x_evm/` and `CONSOLE=tty00` and running `./releasetools/arm_sdimage.sh`. I wrote it to a 16GB SD card with the command: `dd if=minix_arm_sd.img of=/dev/sdb bs=1M oflag=direct`. I put the card in my BeagleBone White Rev A6 and plugged it via USB and started `picocom -b 115200 /dev/ttyUSB0`. It's running the latest HEAD (the version string in the output is 3.4.0), so I didn't mix up the SD cards. I also tried it with another SD card and got the same results.
## Console output

```
...
musb-hdrc: ConfigData=0xde (UTMI-8, dyn FIFOs, HB-ISO Rx, HB-ISO Tx, SoftConn)
musb-hdrc: MHDRC RTL version 2.0 
musb-hdrc: setup fifo_mode 4
musb-hdrc: 28/31 max ep, 16384/16384 memory
USB Peripheral mode controller at 47401000 using PIO, IRQ 0
musb-hdrc: ConfigData=0xde (UTMI-8, dyn FIFOs, HB-ISO Rx, HB-ISO Tx, SoftConn)
musb-hdrc: MHDRC RTL version 2.0 
musb-hdrc: setup fifo_mode 4
musb-hdrc: 28/31 max ep, 16384/16384 memory
USB Host mode controller at 47401800 using PIO, IRQ 0
Net:   <ethaddr> not set. Validating first E-fuse MAC
cpsw, usb_ether
Hit any key to stop autoboot:  0 
mmc0 is current device
SD/MMC found on device 0
reading uEnv.txt
1274 bytes read in 4 ms (310.5 KiB/s)
Loaded environment from uEnv.txt
Importing environment from mmc ...
Running uenvcmd ...
starting from MMC
mmc - MMC sub system

Usage:
mmc read addr blk# cnt
mmc write addr blk# cnt
mmc erase blk# cnt
mmc rescan
mmc part - lists available partition on current mmc device
mmc dev [dev] [part] - show or set current mmc device [partition]
mmc list - lists available devices
reading kernel.bin
264668 bytes read in 35 ms (7.2 MiB/s)
reading ds.elf
79852 bytes read in 13 ms (5.9 MiB/s)
reading rs.elf
109696 bytes read in 17 ms (6.2 MiB/s)
reading pm.elf
81024 bytes read in 13 ms (5.9 MiB/s)
reading sched.elf
44324 bytes read in 9 ms (4.7 MiB/s)
reading vfs.elf
161704 bytes read in 23 ms (6.7 MiB/s)
reading memory.elf
2729976 bytes read in 317 ms (8.2 MiB/s)
reading tty.elf
63388 bytes read in 12 ms (5 MiB/s)
reading mib.elf
78464 bytes read in 13 ms (5.8 MiB/s)
reading vm.elf
165788 bytes read in 23 ms (6.9 MiB/s)
reading pfs.elf
56132 bytes read in 10 ms (5.4 MiB/s)
reading mfs.elf
106172 bytes read in 16 ms (6.3 MiB/s)
reading init.elf
314876 bytes read in 40 ms (7.5 MiB/s)
console=tty00 rootdevname=c0d0p1 verbose=0 hz=1000 board_name=A335BONE
## Starting application at 0x80200000 ...

MINIX 3.4.0. Copyright 2014, Vrije Universiteit, Amsterdam, The Netherlands
MINIX is open source software, see http://www.minix3.org
Started VFS: 9 worker thread(s)
Starting the mmc driver
anon_pagefault: out of memory
VM: pagefault: SIGSEGV 32783 pagefault not handled
PM: coredump signal 11 for 17 / service
service   32783 0x13698 
Out of i-nodes on device 1/6
[1]   Segmentation fault      /bin/service -c ...
anon_pagefault: out of memory
VM: pagefault: SIGSEGV 32784 pagefault not handled
anon_pagefault: out of memory
VM: pagefault: SIGSEGV 32784 pagefault not handled
PM: coredump signal 11 for 18 / init
init*F    32784 0x24d64 
Out of i-nodes on device 1/6
init: single user shell terminated (b), restarting
VM: vm_allocpage: alloc_mem failed
init: can't fork single-user shell: Cannot allocate memory, trying again
VM: vm_allocpage: alloc_mem failed
init: can't fork single-user shell: Cannot allocate memory, trying again
VM: vm_allocpage: alloc_mem failed
...
```
## Build script final output

```
===> Successful make release
===> build.sh ended:      Tue Jan 26 21:08:07 EST 2016
===> Summary of results:
     build.sh command:    build.sh -j 1 -m evbearm-el -O ../obj.evbearm-el -D ../obj.evbearm-el/destdir.evbearm-el -V MKGCCCMDS=yes -V MKLLVM=no -U -u release
     build.sh started:    Tue Jan 26 20:05:39 EST 2016
     MINIX version:       3.4.0
     MACHINE:             evbarm
     MACHINE_ARCH:        earm
     Build platform:      Linux 3.16.0-4-amd64 x86_64
     HOST_SH:             /bin/sh
     /home/tcort/repos/minix/../obj.evbearm-el/tooldir.Linux-3.16.0-4-amd64-x86_64/bin/nbmake outdated (older than usr.bin/make/arch.c), needs building.
     Bootstrapping nbmake
     MAKECONF file:       /etc/mk.conf (File not found)
     TOOLDIR path:        /home/tcort/repos/minix/../obj.evbearm-el/tooldir.Linux-3.16.0-4-amd64-x86_64
     DESTDIR path:        /home/tcort/repos/minix/../obj.evbearm-el/destdir.evbearm-el
     RELEASEDIR path:     /home/tcort/repos/minix/../obj.evbearm-el/releasedir
     Created /home/tcort/repos/minix/../obj.evbearm-el/tooldir.Linux-3.16.0-4-amd64-x86_64/bin/nbmake
     Updated makewrapper: /home/tcort/repos/minix/../obj.evbearm-el/tooldir.Linux-3.16.0-4-amd64-x86_64/bin/nbmake-evbearm-el
     Successful make release
     build.sh ended:      Tue Jan 26 21:08:07 EST 2016
===> .
```

**Comment by dcvmoole** [comment]
That is strange. I hope that this is just a simple bug, because even though a lot of things got bigger since the last release of course, I can hardly believe that that would cause an out-of-memory problem this early in the boot process. We'll definitely have to sort this out before the final 3.4 release, but I for one don't even have a BBW..

**Comment by sambuc** [comment]
I didn't try to run the thing on HW, as the intel memory constraints didn't change so far.

I am building & writing a new SD image, to see if I can reproduce these results as well on a BBW. I'll also check the BBB while I am at it.


Regards,

Lionel

**Comment by sambuc** [comment]
I was able to reproduce the issue, on both BBW & BBB.

I am trying to figure out what is going on.


Regards,

Lionel

**Comment by tcort** [comment]
I tested a bunch of different revisions and found that this issue was introduced in commit 0a6a1f1d05b60e214de2f05a7310ddd1f0e590e7 "NetBSD re-synchronization of the source tree". The commit before it 8933525b85c631b3d20bed940561520be2b81891 "Fix default partition in clientctl." boots OK on my BeagleBone White.

**Comment by pauley-unsaturated** [comment]
I'm having trouble even building those hashes.  Is there a simple way to 'make distclean' or something like that?  I'm afraid my externals are dirty after building with current master.

**Comment by boricj** [comment]
Try `git clean -xdf` and deleting the `obj.evbearm-el` directory. In most cases that should bring back your source tree in a pristine state.

**Comment by pauley-unsaturated** [comment]
I did all of that and I still get:
```
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s: Assembler messages:
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:84: Error: invalid constant (7f) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:104: Error: invalid constant (7f) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:176: Error: invalid constant (ff) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:202: Error: invalid constant (ff) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:212: Error: invalid constant (7f) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:228: Error: invalid constant (7f) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:251: Error: invalid constant (ff) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:272: Error: invalid constant (ff) after mixup
/var/folders/vt/7jjz710n0492gnn2qb99g80c0000gn/T//cc2gNTyo.s:291: Error: invalid constant (ff) after mixup
```

I swear I was building minix-for-arm just a few weeks ago successfully.  Could the underlying gcc version have changed?

**Comment by pauley-unsaturated** [comment]
Never mind, I got it.  Not sure what happened, but I can build with TOT master.  Thanks!

**Comment by pauley-unsaturated** [comment]
I can now repro this as well with a BeagleBone Black. What is the normal procedure for tracking stuff like this down? JTAG?

**Comment by tcort** [comment]
> What is the normal procedure for tracking stuff like this down?

Manually inspecting the contents of the commit that caused the breakage to look for clues. Unfortunately, the commit causing the problem involves over 32,000 files.

**Comment by pauley-unsaturated** [comment]
32k files sounds a bit rough without being able to narrow down exactly what is causing this.  If we had a back-trace we could at least know which source files to check.  Is there a mechanism to print the symbolicated backtrace here, in servers/vm?

The good thing is that we can guess this is limited to ARM.  The BeagleBoneBlack has 512MB of memory, so it can't be that we're running out. This happens right after the mmc server comes online.  The call in question that's failing takes an alignment flag.  I'm going to add a bit of logging to see if we're able to allocate _any_ memory at all, and at what alignment values.. but gosh it takes a while to flash an SD card. I may need to try net-booting as your article specifies, @tcort.

**Comment by pauley-unsaturated** [comment]
It looks like memory is being allocated, and we are actually running through it.  Maybe this has to do with something that changed in the several things inside of common/lib/libc/arch/arm.  I'm going to attempt bisecting the arm related changes by rolling back to the last working commit.

**Comment by pauley-unsaturated** [comment]
Rolling back libc doesn't seem to change anything.  I'm going to try rolling back the bsd compiler_rt changes now which also have a bunch of arm specific changes.

**Comment by pauley-unsaturated** [comment]
Rolling back the changes in both of the vmparam.h inside of /sys/arch/arm for the diff in question has no effect.

**Comment by boricj** [comment]
I just rebased the current work-in-progress port to Raspberry Pi 2 and 3 on master to see what happens. With almost 1 GiB of RAM MINIX doesn't boot on the real hardware (it just hangs after the copyright banner) but it boots on QEMU.

Of course I can't rule out an issue with the Raspberry Pi port itself, but I noticed while porting that the micro-kernel is hard-coded to use 256 MiB of RAM on ARM (in `minix/kernel/arch/earm/pre_init.c`), even though the BBB has 512 MiB of RAM. Perhaps taking advantage of the extra RAM would make MINIX boot once again on the BBB ?

**Comment by pauley-unsaturated** [comment]
Wow, that's a good find. Have you diffed that file against the previous working version for BBB? I might try doing so today, since it's a holiday on the states.


-Pauley

**Comment by boricj** [comment]
`git diff 8933525..0a6a1f1 -- minix/kernel/arch/earm/pre_init.c` is empty.

I don't own a Beagle{Board,Bone} so unfortunately I can't debug this on my side. If you increase the amount of RAM declared, you'll have to modify `setup_mbi()` in `minix/kernel/arch/earm/pre_init.c` and possibly `pg_identity()` in `minix/kernel/arch/earm/pg_utils.c`.

**Comment by dcvmoole** [comment]
Fortunately, the amount of memory is not the problem here (I say "fortunately" in the sense that minix isn't *that* memory hungry yet hehe): so far we do know that something is causing one of the early processes to eat up all the system memory, no matter how much memory the system has. The last I heard regarding this issue here is that there might be a problem with the gcc crosscompiler but that is still conjecture. Whether that is in any way related to what you're seeing on the RPi hardware @boricj is anybody's guess (awesome work btw!!), but it does look like at least the symptoms are different.

The reason for the 256mb is that the beaglebone black and white use the same image, and that image is therefore hardcoded to the smallest amount of memory between the two, which is the BBW's 256mb. That is of course highly undesirable and should be fixed, but that really is another issue.

**Comment by skilbjo** [comment]
Just bought a Beaglebone Black hoping to load minix on it. Checked out master and followed the build steps but I get the errors referenced in the issue above. Safe to say even though BBB is "officially" supported, the code in master doesn't work on ARM platforms?

**Comment by sambuc** [comment]
As you can see this issue is a release blocker.

Any help is appreciated, work has progressed but is not yet complete enough to be merged in master, see #195 .

If you wish to simply use minix, you can still use the commit before the update of the gcc toolchain.

**Comment by skilbjo** [comment]
Gotcha. Yeah I'm excited about the invitation for contribution, but looking through #195 , C/C++ and clang/gcc toolchains and the Minix project directory structure are really not my skillset at the moment... :(  Unless there are user meetups in SF? That would be a good way to ask the experts how to get started on the project

I did do `git reset --hard 8933525b85c631b3d20bed940561520be2b81891` and `./releasetools/arm_sdimage.sh` to try to get Minix on BBB but got back on the build
    
    ERROR: Failed to make release
    *** BUILD ABORTED ***

so not really sure. I'll just cross my fingers and send you best wishes from userland on work to be done for #195

**Comment by sambuc** [comment]
Did you remove the `obj.evbearm` folder, before re-running the `arm_sdimage.sh` script?

While you are at it, also make sure to define the `.settings` file, or to set the variables as required for a BBB build.

**Comment by skilbjo** [comment]
Sorry for the nasty stack trace. Yeah, it seems not not be building back on that commit :/

I'm on macOS Sierra 10.12.5 using the latest software updates. 

```
@mbp:minix $ cc --version
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

```
@mbp:stuff $ git clone git@github.com:Stichting-MINIX-Research-Foundation/minix.git
Cloning into 'minix'...
remote: Counting objects: 184389, done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 184389 (delta 0), reused 1 (delta 0), pack-reused 184382
Receiving objects: 100% (184389/184389), 177.07 MiB | 667.00 KiB/s, done.
Resolving deltas: 100% (93149/93149), done.
Checking out files: 100% (71048/71048), done.
@mbp:stuff $ cd minix/
@mbp:minix $ git reset --hard 8933525b85c631b3d20bed940561520be2b81891
Checking out files: 100% (41677/41677), done.
HEAD is now at 8933525b8 Fix default partition in clientctl.
@mbp:minix $ touch .settings
@mbp:minix $ vim .settings
@mbp:minix $ ./releasetools/arm_sdimage.sh
Sourcing settings from .settings
CONTENT #beaglebone (and black)
CONTENT U_BOOT_BIN_DIR=build/am335x_evm/
CONTENT CONSOLE=tty00
Going to build Minix source code...
===> build.sh command:    build.sh -j 1 -m evbearm-el -O ../obj.evbearm-el -D ../obj.evbearm-el/destdir.evbearm-el -V MKGCCCMDS=yes -V MKLLVM=no -U -u release
===> build.sh started:    Sun Jun 11 18:31:29 PDT 2017
===> MINIX version:       3.3.0
===> MACHINE:             evbarm
===> MACHINE_ARCH:        earm
===> Build platform:      Darwin 16.6.0 x86_64
===> HOST_SH:             /bin/sh
===> No $TOOLDIR/bin/nbmake, needs building.
===> Bootstrapping nbmake
...etc...
/Users/skilbjo/dev/stuff/minix/../obj.evbearm-el/tooldir.Darwin-16.6.0-x86_64/bin/arm-elf32-minix-gcc -O2 -std=gnu99 -Werror   -msoft-float   --sysroot=/Users/skilbjo/dev/stuff/minix/../obj.evbearm-el/destdir.evbearm-el -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc/libgcc -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc/arch/earm -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/usr.bin/gcc/arch/earm -DIN_GCC -DHAVE_GTHR_DEFAULT -DIN_LIBGCC2 -D__GCC_FLOAT_NOT_NEEDED -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/. -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/../include -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/../libcpp/include -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/../libdecnumber -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/../libdecnumber/dpd  -fPIC -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/config/arm -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/ginclude -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/cp -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/dist/gcc/config -I/Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libstdc++-v3/arch/earm -I.  -c   _popcountsi2.c -o _popcountsi2.o.tmp1
/var/folders/lg/rk5w7m3j2ls3gw1hynzl18k40000gn/T//ccRuyeaY.s: Assembler messages:
/var/folders/lg/rk5w7m3j2ls3gw1hynzl18k40000gn/T//ccRuyeaY.s:33: Error: invalid constant (ff) after fixup
*** [_popcountsi2.o] Error code 1

nbmake[8]: stopped in /Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc/libgcc
1 error

nbmake[8]: stopped in /Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc/libgcc
*** [dependall] Error code 2

nbmake[7]: stopped in /Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc/libgcc
1 error

nbmake[7]: stopped in /Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc/libgcc
*** [dependall-libgcc] Error code 2

nbmake[6]: stopped in /Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc
1 error

nbmake[6]: stopped in /Users/skilbjo/dev/stuff/minix/external/gpl3/gcc/lib/libgcc

*** Failed target:  dependall-../external/gpl3/gcc/lib/libgcc
*** Failed command: _makedirtarget() { dir="$1"; shift; target="$1"; shift; case "${dir}" in /*) this="${dir}/"; real="${dir}" ;; .) this="lib/"; real="/Users/skilbjo/dev/stuff/minix/lib" ;; *) this="lib/${dir}/"; real="/Users/skilbjo/dev/stuff/minix/lib/${dir}" ;; esac; show=${this:-.}; echo "${target} ===> ${show%/}${1:+ (with: $@)}"; cd "${real}" && /Users/skilbjo/dev/stuff/minix/../obj.evbearm-el/tooldir.Darwin-16.6.0-x86_64/bin/nbmake _THISDIR_="${this}" "$@" ${target}; }; _makedirtarget ../external/gpl3/gcc/lib/libgcc dependall
*** Error code 2

Stop.
nbmake[5]: stopped in /Users/skilbjo/dev/stuff/minix/lib
*** [build_install] Error code 1

nbmake[4]: stopped in /Users/skilbjo/dev/stuff/minix/lib
1 error

nbmake[4]: stopped in /Users/skilbjo/dev/stuff/minix/lib
*** [do-lib] Error code 2

nbmake[3]: stopped in /Users/skilbjo/dev/stuff/minix
1 error

nbmake[3]: stopped in /Users/skilbjo/dev/stuff/minix
*** [build] Error code 2

nbmake[2]: stopped in /Users/skilbjo/dev/stuff/minix
1 error

nbmake[2]: stopped in /Users/skilbjo/dev/stuff/minix
*** [distribution] Error code 2

nbmake[1]: stopped in /Users/skilbjo/dev/stuff/minix
1 error

nbmake[1]: stopped in /Users/skilbjo/dev/stuff/minix
*** [release] Error code 2

nbmake: stopped in /Users/skilbjo/dev/stuff/minix
1 error

nbmake: stopped in /Users/skilbjo/dev/stuff/minix

ERROR: Failed to make release
*** BUILD ABORTED ***
```

**Comment by rvalles** [comment]
What's the status of this?

**Comment by stux2000** [comment]
Since this issue is closed, what is the status on the 3.4.0 release?  Are we waiting on the clang compile fix (pull request #105) or will that be pushed for a 3.5.0 release? (I'm not sure what the best place to ask is.)
